internal/ingress/controller/template/template.go

```diff
@@ -1013,13 +1014,39 @@ func buildOpentracing(c interface{}, s interface{}) string {
 	}
 
 	buf := bytes.NewBufferString("")
+	var host string = "localhost"
 
 	if cfg.DatadogCollectorHost != "" {
 		buf.WriteString("opentracing_load_tracer /usr/local/lib64/libdd_opentracing.so /etc/nginx/opentracing.json;")
+		host = cfg.DatadogCollectorHost
 	} else if cfg.ZipkinCollectorHost != "" {
 		buf.WriteString("opentracing_load_tracer /usr/local/lib/libzipkin_opentracing_plugin.so /etc/nginx/opentracing.json;")
-	} else if cfg.JaegerCollectorHost != "" || cfg.JaegerEndpoint != "" {
+		host = cfg.ZipkinCollectorHost
+	} else if cfg.JaegerEndpoint != "" { // Takes priority over the collector within the cpp client code
+		// from https://github.com/jaegertracing/jaeger-client-cpp/blob/c1d57957e8478667ba8d46cba7fa6ad2dbd33f85/src/jaegertracing/net/URI.cpp#L67
+		jaegerRegexp := regexp.MustCompile(`^(([^:\/?#]+):)?(\/\/([^\/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?`)
+		hostMatch := jaegerRegexp.FindStringSubmatch(cfg.JaegerEndpoint)
+		if hostMatch == nil {
+			klog.Errorf("could not find a valid hostname in cfg.JaegerEndpoint (%v)", cfg.JaegerEndpoint)
+			return ""
+		}
+		buf.WriteString("opentracing_load_tracer /usr/local/lib/libjaegertracing_plugin.so /etc/nginx/opentracing.json;")
+		host = strings.Split(hostMatch[4], ":")[0]
```

Comment:
the above logic is taken from the [client code](https://github.com/jaegertracing/jaeger-client-cpp/blob/c1d57957e8478667ba8d46cba7fa6ad2dbd33f85/src/jaegertracing/net/URI.cpp#L67). I'm keeping the regex confined to `cfg.JaegerEndpoint` because that regex expects something more specific. For example, just saying `mydomain.com` (for Zipkin, let's say) would actually result in regex match group 5 (not 4) but should still be valid. Meeting Jaeger endpoint standards simplifies the logic required in ingress-nginx, but I still do the split on `://` afterward in case another configuration included a protocol.